### PR TITLE
fleetctl: check also systemd states after starting units

### PR DESCRIPTION
**Pull request by dongsupark**
To resolve inconsistency between ``"systemctl start"`` and ``"fleetctl start"``, fleetctl needs to check systemd states after starting units. If the systemd state is not a desired one, wait until it gets an active state.

Fixes: https://github.com/coreos/fleet/issues/1025

### CODE REVIEW

**Comment by dongsupark** [comment]
Updated. (after several hiccups)
* Handle return value of ``tryWaitForSystemdActiveState`` carefully. It should not return 1 only because of a missing systemd dbus connection. Otherwise it will break unit tests based on Travis, as that environment cannot correctly set up a systemd dbus connection.
* Added a missing nil checking before ``conn.Close()``. Also call ``conn.Close()`` after checking for nil.
* Wrote comments etc.

**Comment by dongsupark** [comment]
I'm going to merge it soon, unless anyone has other opinions.

File: fleetctl/fleetctl.go [hunk 1]
```diff
@@ -1060,6 +1061,87 @@ func assertUnitState(name string, js job.JobState, out io.Writer) (ret bool) {
...
+// active state, making use of the systemd's DBUS interface. First it takes one
+// or more units as input, and ensures that every unit in the []units must be
+// in the active state. If yes, return 0. Otherwise return 1.
+//
+// NOTE: To avoid unnecessary conflicts with unit tests in start_test.go, we
+// should return 0 for the case of sd_dbus.New() returning error. In that case,
+// it just means that the unit test environment based on Travis is not capable
+// of setting up a DBUS connection of systemd, which would actually work fine
+// for production systems or functional tests.
+func tryWaitForSystemdActiveState(units []string) (ret int) {
```
Comment (jonboulle) [inline fleetctl/fleetctl.go]: imho the ret value is pretty unidiomatic; rather, return an error, and have the callers decide to log the error or not

File: fleetctl/fleetctl.go [hunk 2]
```diff
@@ -1060,6 +1061,87 @@ func assertUnitState(name string, js job.JobState, out io.Writer) (ret bool) {
...
+	conn, err := sd_dbus.New()
+	if err != nil {
+		stderr("Failed to get systemd-dbus conn: %v", err)
+		return 0
+	}
+	if conn == nil {
```
Comment (jonboulle) [inline fleetctl/fleetctl.go]: can this happen?

File: fleetctl/fleetctl.go [hunk 3]
```diff
@@ -1060,6 +1061,87 @@ func assertUnitState(name string, js job.JobState, out io.Writer) (ret bool) {
...
+	for attempt := 0; attempt < maxAttempts; attempt++ {
+		found, errAssert = assertSystemdActiveState(conn, unitName)
+		if found && errAssert == nil {
+			break
+		}
+		time.Sleep(sleep)
+	}
+
+	if found {
+		err = nil
```
Comment (jonboulle) [inline fleetctl/fleetctl.go]: unnecessary

File: fleetctl/fleetctl.go [hunk 4]
```diff
@@ -1060,6 +1061,87 @@ func assertUnitState(name string, js job.JobState, out io.Writer) (ret bool) {
...
+// assertSystemdActiveState determines if a given systemd unit is actually
+// in the active state, making use of systemd's DBUS connection.
+func assertSystemdActiveState(conn *sd_dbus.Conn, unitName string) (found bool, err error) {
+	listUnits, err := conn.ListUnits()
+	if err != nil {
+		return false, fmt.Errorf("Failed to list systemd unit: %v", err)
+	}
+
+	for _, u := range listUnits {
+		if strings.Compare(u.Name, unitName) != 0 {
```
Comment (jonboulle) [inline fleetctl/fleetctl.go]: == works fine for strings

File: fleetctl/start.go [hunk 5]
```diff
@@ -87,5 +87,10 @@ func runStartUnit(cCmd *cobra.Command, args []string) (exit int) {
 		return exitVal
 	}
 
+	exitVal = tryWaitForSystemdActiveState(starting)
```
Comment (jonboulle) [inline fleetctl/start.go]: This needs to consider block-attempts and no-block flags right?

**Comment by jonboulle** [comment]
I think this is wrong; it's checking against a local systemd instance over dbus, but that will ONLY be right if `fleetctl` happens to be invoked on the same system on which the units are scheduled, right? Surely we need to check against the API, as we do with the other states?

File: fleetctl/fleetctl.go (hunk 1 above)
Comment (dongsupark) [inline fleetctl/fleetctl.go]: Agreed. I'll try to make it return an error.

File: fleetctl/fleetctl.go (hunk 3 above)
Comment (dongsupark) [inline fleetctl/fleetctl.go]: Correct. I'll just remove `err = nil`.

File: fleetctl/fleetctl.go (hunk 4 above)
Comment (dongsupark) [inline fleetctl/fleetctl.go]: Right. I'll just use `==`.

File: fleetctl/start.go (hunk 5 above)
Comment (dongsupark) [inline fleetctl/start.go]: Geez, you're right. I'll pass `getBlockAttempts(cCmd)` to that function to check the flags, just like `tryWaitForUnitStates()`.

**Comment by dongsupark** [comment]
@jonboulle Ok. I'll try to take a different approach as you suggested. From my understanding, `schema.UnitState` has already its own fields like `SystemdActiveState`, `SystemdLoadState`, and `SystemdSubState`. So we should make use of `cAPI.UnitStates()` to retrieve these states, to make sure that these are the expected states. Is that correct?

**Comment by jonboulle** [comment]
Yep, that sounds about right!

**Pull request merged**